src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/resource_allocation_manager_impl.cc

```diff
@@ -0,0 +1,506 @@
+#include "application_manager/application.h"
+#include "application_manager/application_manager.h"
+#include "interfaces/HMI_API.h"
+#include "interfaces/MOBILE_API.h"
+#include "smart_objects/enum_schema_item.h"
+#include "rc_rpc_plugin/rc_rpc_plugin.h"
+#include "application_manager/message_helper.h"
+#include "rc_rpc_plugin/rc_module_constants.h"
+#include "json/json.h"
+#include "utils/helpers.h"
+#include "utils/make_shared.h"
+
+//#include "remote_control/message_helper.h"
+//#include "remote_control/rc_app_extension.h"
+
+namespace rc_rpc_plugin {
+
+CREATE_LOGGERPTR_GLOBAL(logger_, "RemoteControlModule")
+
+ResourceAllocationManagerImpl::ResourceAllocationManagerImpl(
+    application_manager::ApplicationManager& app_mngr,
+    application_manager::rpc_service::RPCService& rpc_service)
+    : current_access_mode_(hmi_apis::Common_RCAccessMode::AUTO_ALLOW)
+    , app_mngr_(app_mngr)
+    , rpc_service_(rpc_service)
+    , is_rc_enabled_(true) {}
+
+ResourceAllocationManagerImpl::~ResourceAllocationManagerImpl() {}
+
+AcquireResult::eType ResourceAllocationManagerImpl::AcquireResource(
+    const std::string& module_type, const uint32_t app_id) {
+  LOG4CXX_AUTO_TRACE(logger_);
+  const application_manager::ApplicationSharedPtr acquiring_app =
+      app_mngr_.application(app_id);
+  if (!acquiring_app) {
+    LOG4CXX_WARN(logger_, "App with app_id: " << app_id << "does not exist!");
+    return AcquireResult::IN_USE;
+  }
+
+  sync_primitives::AutoLock lock(allocated_resources_lock_);
+  const AllocatedResources::const_iterator allocated_it =
+      allocated_resources_.find(module_type);
+  if (allocated_resources_.end() == allocated_it) {
+    SetResourceAquired(module_type, app_id);
+    LOG4CXX_DEBUG(logger_,
+                  "Resource is not acquired yet. "
+                      << "App: " << app_id << " is allowed to acquire "
+                      << module_type);
+    return AcquireResult::ALLOWED;
+  }
+
+  if (app_id == allocated_resources_[module_type]) {
+    LOG4CXX_DEBUG(logger_,
+                  "App: " << app_id << " is already acquired resource "
+                          << module_type);
+    return AcquireResult::ALLOWED;
+  }
+
+  if (IsModuleTypeRejected(module_type, app_id)) {
+    LOG4CXX_DEBUG(logger_,
+                  "Driver disallowed app: " << app_id << " to acquire "
+                                            << module_type);
+    return AcquireResult::REJECTED;
+  }
+
+  const mobile_apis::HMILevel::eType acquiring_app_hmi_level =
+      acquiring_app->hmi_level();
+
+  if (mobile_apis::HMILevel::HMI_FULL != acquiring_app_hmi_level) {
+    LOG4CXX_DEBUG(logger_,
+                  "Aquiring resources is not allowed in HMI level: "
+                      << acquiring_app_hmi_level << ". App: " << app_id
+                      << " is disallowed to acquire " << module_type);
+    return AcquireResult::REJECTED;
+  }
+
+  switch (current_access_mode_) {
```

Comment:
I do not see the code of handling the case that 
the resource is allocate to one app, and there is a pending request (for example SetInteriorVechileData or BUttonPressed) waiting for HMI side to finish (an example is that in order to set fan speed to 80% from 0%, the system may need to increase the fan speed step by step internally, thus it cannot be finished within a very short time (<100ms) )
if we assume the HMI side cannot handle multiple such requests at the same time, SDL shall reject the new request from another app and not allocate the resource to the new app.

Reply:
Requesting Application state | Requested module status | RC setting - Access mode | Expected SDL action
-- | -- | -- | --
any | busy | any | disallow